### PR TITLE
Improve ProxyAuth

### LINKED ISSUE #1722: Proxy-Authentication does not work with HTTPS sites
##### Steps to reproduce the problem:

1. `mitmdump --nonanonymous`
2. `curl -x localhost:8080 -k -U foo:bar http://example.com` (works)
3. `curl -x localhost:8080 -k -U foo:bar https://example.com` (doesn't)

##### Any other comments? What have you tried so far?

The problem is an invalid handling of `http_authenticated`.

### PR DESCRIPTION
**Pull request by mhils**
This PR fixes #1722 and #1734, and hopefully #2019 as well. 
@laurmurclar, can you check if that works for you?

### CODE REVIEW

**Comment by laurmurclar** [comment]
The dialog requesting credentials for the proxy no longer shows up, but I'm still getting the ERR_SPDY_PROTOCOL_ERROR problem and having to give credentials for HTTPS sites that don't usually need any. 😕 Thanks for working on this so quickly!

**Comment by laurmurclar** [comment]
It does 🎉 I was just being an idiot earlier, sorry 😬 thanks again for fixing this so quickly

**Comment by mhils** [comment]
No problem, thanks for testing that so quickly!

File: mitmproxy/addons/proxyauth.py [hunk 1]
```diff
@@ -1,103 +1,117 @@
...
+    return scheme, user, password
 
 
 class ProxyAuth:
     def __init__(self):
         self.nonanonymous = False
         self.htpasswd = None
         self.singleuser = None
+        self.mode = None
+        self.authenticated = weakref.WeakSet()  # type: Set[connections.ClientConnection]
```
Comment (Kriechi) [inline mitmproxy/addons/proxyauth.py]: Is this voodoo black magic really necessary?

File: mitmproxy/addons/proxyauth.py [hunk 2]
```diff
@@ -1,103 +1,117 @@
...
+                "Proxy Authentication Required",
+                mitmproxy.net.http.Headers(Proxy_Authenticate='Basic realm="{}"'.format(REALM)),
             )
         else:
             return http.make_error_response(
-                407,
-                "Proxy Authentication Required",
-                headers,
+                401,
+                "Authentication Required",
```
Comment (Kriechi) [inline mitmproxy/addons/proxyauth.py]: Please use the already defined reasons: https://github.com/mitmproxy/mitmproxy/blob/master/mitmproxy/net/http/status_codes.py#L82

File: mitmproxy/addons/proxyauth.py [hunk 3]
```diff
@@ -1,103 +1,117 @@
...
+        except ValueError:
+            return None
 
         if self.nonanonymous:
-            pass
+            return username, password
         elif self.singleuser:
-            if [username, password] != self.singleuser:
-                return False
+            if [username, password] == self.singleuser:
```
Comment (Kriechi) [inline mitmproxy/addons/proxyauth.py]: Is this a [Yoda condition](https://en.wikipedia.org/wiki/Yoda_conditions)?

**Pull request merged**